lib/srv/discovery/fetchers/eks.go

```diff
@@ -319,5 +354,336 @@ func (a *eksFetcher) getMatchingKubeCluster(ctx context.Context, clusterName str
+			TLSClientConfig: rest.TLSClientConfig{
+				CAData: ca,
+			},
+		},
+	)
+	return client, trace.Wrap(err, "unable to create Kubernetes client for cluster %q", aws.StringValue(cluster.Name))
+}
+
+// upsertClusterRoleWithAdminCredentials tries to upsert the ClusterRole using admin credentials.
+func (a *eksFetcher) upsertClusterRoleWithAdminCredentials(ctx context.Context, client *kubernetes.Clientset) error {
+	clusterRole := &rbacv1.ClusterRole{}
+
+	if err := yaml.Unmarshal([]byte(fixtures.KubeClusterRoleTemplate), clusterRole); err != nil {
+		return trace.Wrap(err)
+	}
+
+	_, err := client.RbacV1().ClusterRoles().Create(ctx, clusterRole, metav1.CreateOptions{})
+	if err == nil {
+		return nil
+	}
+
+	if kubeerrors.IsAlreadyExists(err) {
+		_, err := client.RbacV1().ClusterRoles().Update(ctx, clusterRole, metav1.UpdateOptions{})
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(err)
+}
+
+// upsertClusterRoleBindingWithAdminCredentials tries to upsert the ClusterRoleBinding using admin credentials
+// and maps it into the principal group.
+func (a *eksFetcher) upsertClusterRoleBindingWithAdminCredentials(ctx context.Context, client *kubernetes.Clientset, groupID string) error {
+	clusterRoleBinding := &rbacv1.ClusterRoleBinding{}
+
+	if err := yaml.Unmarshal([]byte(fixtures.KubeClusterRoleBindingTemplate), clusterRoleBinding); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if len(clusterRoleBinding.Subjects) == 0 {
+		return trace.BadParameter("Subjects field were not correctly unmarshaled")
+	}
+
+	clusterRoleBinding.Subjects[0].Name = groupID
+
+	_, err := client.RbacV1().ClusterRoleBindings().Create(ctx, clusterRoleBinding, metav1.CreateOptions{})
+	if err == nil {
+		return nil
+	}
+
+	if kubeerrors.IsAlreadyExists(err) {
+		_, err := client.RbacV1().ClusterRoleBindings().Update(ctx, clusterRoleBinding, metav1.UpdateOptions{})
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(err)
+}
+
+// upsertAccessEntry upserts the access entry for the specified ARN with the teleportKubernetesGroup.
+func (a *eksFetcher) upsertAccessEntry(ctx context.Context, client eksiface.EKSAPI, cluster *eks.Cluster) error {
+	_, err := convertAWSError(
+		client.CreateAccessEntryWithContext(ctx,
+			&eks.CreateAccessEntryInput{
+				ClusterName:      cluster.Name,
+				PrincipalArn:     aws.String(a.SetupAccessForARN),
+				KubernetesGroups: aws.StringSlice([]string{teleportKubernetesGroup}),
+			},
+		))
+	if err == nil || !trace.IsAlreadyExists(err) {
+		return trace.Wrap(err)
+	}
+
+	_, err = convertAWSError(
+		client.UpdateAccessEntryWithContext(ctx,
+			&eks.UpdateAccessEntryInput{
+				ClusterName:      cluster.Name,
+				PrincipalArn:     aws.String(a.SetupAccessForARN),
+				KubernetesGroups: aws.StringSlice([]string{teleportKubernetesGroup}),
+			},
+		))
+
+	return trace.Wrap(err)
+}
+
+func (a *eksFetcher) setCallerIdentity(ctx context.Context) error {
+	if a.AssumeRole.RoleARN != "" {
+		a.callerIdentity = a.AssumeRole.RoleARN
+	}
```

Comment:
Why do we need this? We reassign it at the end (at least in a happy path, but I don't see a reason for setting it for error case)

Reply:
I missed the return nil